src/Chain.cpp

```diff
@@ -112,15 +112,55 @@ std::vector<int> Chain::ViterbiPath(int row, int col) const {
 /// column.
 IntVectorVector Chain::ViterbiPaths() const {
   IntVectorVector paths;
-  for (int i = 0; i < viterbi_idx_.rows(); i++) {
-    for (int j = 0; j < viterbi_idx_.cols(); j++) {
+  // Note that we call `viterbi_idx()` (not `viterbi_idx_`) because this ensures
+  // `viterbi_idx_` is computed.
+  for (int i = 0; i < viterbi_idx().rows(); i++) {
+    for (int j = 0; j < viterbi_idx().cols(); j++) {
       paths.push_back(ViterbiPath(i, j));
     }
   }
   return paths;
 };
 
 
+/// @brief If a Chain is clean, mark it as dirty and recursively mark the
+/// downstream Smooshishs as dirty.
+void Chain::MarkAsDirty() {
+  if (!is_dirty_) {
+    is_dirty_ = true;
+    prev_->MarkAsDirty();
+    curr_->MarkAsDirty();
+  }
+};
+
+
+/// @brief If a Chain is dirty, mark it as clean, resize its marginal/viterbi
+/// probability matrices to empty matrices, and recursively mark the downstream
+/// Smooshishs as clean.
```

Comment:
Aha! If only I could remember things.

Please note that setting the matrices to empty will force recomputation.